xdsl/transforms/csl_stencil_to_csl_wrapper.py

```diff
@@ -0,0 +1,232 @@
+        width: int = 1
+        height: int = 1
+        z_dim_no_ghost_cells: int = 1
+        z_dim: int = 1
+        for apply_op in apply_ops:
+            # loop over accesses to get `neighbour` (from which we build `pattern`)
+            for ap in apply_op.get_accesses():
+                if ap.is_diagonal:
+                    raise ValueError("Diagonal accesses are currently not supported")
+                # if ap.dims != 2:
+                #     raise ValueError("Stencil accesses must be 2-dimensional at this stage")
+                if len(ap.offsets) > 0:
+                    neighbours = max(
+                        neighbours, max(abs(d) for offset in ap.offsets for d in offset)
+                    )
+
+            # find max x and y dimensions
+            if len(shape := apply_op.topo.shape.data) == 2:
+                assert isinstance(
+                    shape.data[0].data, int
+                ), "Cannot have a float data shape"
+                assert isinstance(
+                    shape.data[1].data, int
+                ), "Cannot have a float data shape"
+                width = max(width, shape.data[0].data)
+                height = max(height, shape.data[1].data)
+            else:
+                raise ValueError("Stencil accesses must be 2-dimensional at this stage")
+
+            # find max z dimension - we could get this from func args, store ops, or apply ops
+            for result in apply_op.results:
+                if isa(result.type, stencil.TempType[TensorType[Attribute]]):
+                    z_dim_no_ghost_cells = max(
+                        z_dim_no_ghost_cells,
+                        result.type.get_element_type().get_shape()[0],
+                    )
+            for arg in op.args:
+                if isa(field_t := arg.type, stencil.FieldType[TensorType[Attribute]]):
+                    z_dim = max(z_dim, field_t.get_element_type().get_shape()[0])
+
+        # initialise module op
+        module_op = csl_wrapper.ModuleOp(
+            width=IntegerAttr(width, 16),
+            height=IntegerAttr(height, 16),
+            params={
+                "z_dim": IntegerAttr(z_dim, 16),
+                "pattern": IntegerAttr(neighbours + 1, 16),
+            },
+        )
+
+        self.initialise_layout_module(module_op)
+        module_op.update_program_block_args_from_layout()
+        module_op.set_program_name(op.sym_name)
+
+        # add func args to program_module block args
+        for block_arg in op.body.block.args:
+            module_op.program_module.block.insert_arg(
+                block_arg.type, len(module_op.program_module.block.args)
+            )
+
+        # replace func.return
+        func_return = op.body.block.last_op
+        assert isinstance(func_return, func.Return)
+        assert (
+            len(func_return.arguments) == 0
+        ), "Non-empty returns currently not supported"
+        rewriter.replace_op(func_return, csl.ReturnOp())
```

Comment:
How would we handle return values returned by the function?

Reply:
By the above assert 🤓

Reply:
I'm actually not sure how we could export un-named return values for host-device transfers.